File: leeetcode/2020-Challenges/04/week 1/D3.MaximumSubarray.swift
/*
Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.

Example:

Input: [-2,1,-3,4,-1,2,1,-5,4],
Output: 6
Explanation: [4,-1,2,1] has the largest sum = 6.
Follow up:

If you have figured out the O(n) solution, try coding another solution using the divide and conquer approach, which is more subtle.
*/
class Solution {
    func maxSubArray(_ nums: [Int]) -> Int {
        var max = Int.min
        var current = 0
        
        for i in 0..<nums.count{
            current += nums[i]
            if(current > max){
                max = current
            }
            if(current < 0){
                current = 0
            }
        }
        
        return max
    }
}

File: leeetcode/2020-Challenges/04/week 1/D3.MaximumSubarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int max = Integer.MIN_VALUE, candidate = 0;
        if(nums == null) return -1;
        for(int i = 0; i < nums.length; i++){
            candidate += nums[i];
            if(candidate > max){
                max = candidate;
            }
            
            if(candidate < 0){
                candidate = 0;
            }
        }
    return max;
    }
}